crates/oxc_mangler/src/lib.rs

```diff
@@ -125,59 +128,80 @@ impl Mangler {
 
         // Walk the scope tree and compute the slot number for each scope
         let mut tmp_bindings = std::vec::Vec::with_capacity(100);
-        for scope_id in scope_tree.descendants_from_root() {
+        for scope_id in iter::once(scope_tree.root_scope_id())
+            .chain(scope_tree.iter_all_child_ids(scope_tree.root_scope_id()))
+        {
             let bindings = scope_tree.get_bindings(scope_id);
-
-            // The current slot number is continued by the maximum slot from the parent scope
-            let parent_max_slot = scope_tree
-                .get_parent_id(scope_id)
-                .map_or(0, |parent_scope_id| max_slot_for_scope[parent_scope_id.index()]);
-
-            let mut slot = parent_max_slot;
-
-            if !bindings.is_empty() {
-                // Sort `bindings` in declaration order.
-                tmp_bindings.clear();
-                tmp_bindings.extend(bindings.values().copied());
-                tmp_bindings.sort_unstable();
-                for symbol_id in &tmp_bindings {
-                    slots[symbol_id.index()] = slot;
-                    slot += 1;
-                }
+            if bindings.is_empty() {
+                continue;
             }
 
-            max_slot_for_scope[scope_id.index()] = slot;
+            let mut slot = slot_liveness.len();
+
+            let nearest_var_scope_id = scope_tree
+                .ancestors(scope_id)
+                .find(|s_id| scope_tree.get_flags(*s_id).is_var())
+                .unwrap_or(scope_tree.root_scope_id());
+
+            let reusable_slots = Vec::from_iter_in(
+                slot_liveness
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, slot_liveness)| !slot_liveness.contains(scope_id.index()))
+                    .map(|(slot, _)| slot)
+                    .take(bindings.len()),
+                &allocator,
+            );
+
+            let remaining_count = bindings.len() - reusable_slots.len();
+
+            let assignable_slots = reusable_slots.into_iter().chain(slot..slot + remaining_count);
+            slot += remaining_count;
+            if slot_liveness.len() < slot {
+                slot_liveness.resize_with(slot, || FixedBitSet::with_capacity(scope_tree.len()));
+            }
 
-            if slot > total_number_of_slots {
-                total_number_of_slots = slot;
+            // Sort `bindings` in declaration order.
+            tmp_bindings.clear();
+            tmp_bindings.extend(bindings.values().copied());
+            tmp_bindings.sort_unstable();
+            for (symbol_id, assigned_slot) in tmp_bindings.iter().zip(assignable_slots) {
+                slots[symbol_id.index()] = assigned_slot;
+
+                let lived_scope_ids = symbol_table
+                    .get_resolved_references(*symbol_id)
+                    .flat_map(|reference| {
+                        let scope_id = ast_nodes.get_node(reference.node_id()).scope_id();
```

Comment:
I fixed the previous attempt by going through the nodes table ... @overlookmotel please don't hate me :-)